Windows setting language

Gijsreyn opened the pull request:
<!-- To check a checkbox place an "x" between the brackets. e.g: [x] -->

- [x] I have signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-dsc).
- [x] This pull request is related to an issue.

-----

Addresses issue #34 and #92. Only the module this time.

ryfu-msft commented on resources/Help/Microsoft.Windows.Setting.Language/DisplayLanguage.md:
```diff
@@ -0,0 +1,36 @@
+
+| **Parameter** | **Attribute** | **DataType** | **Description**                                                                                                                           | **Allowed Values**                                                               |
+| ------------- | ------------- | ------------ | ----------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------- |
+| `LocaleName`  | Mandatory     | String       | The name of the language. This is the language tag that represents the language. For example, `en-US` represents English (United States). | Use the `Get-WinUserLanguageList` to see what language pack have been installed. |
```
```suggestion
| `LocaleName`  | Mandatory     | String       | The name of the language. This is the language tag that represents the language. For example, `en-US` represents English (United States). | Use the `Get-WinUserLanguageList` to see which language packs have been installed. |
```

ryfu-msft commented on resources/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.psm1:
```diff
@@ -0,0 +1,213 @@
+}
+
+#region Functions
+function Get-OsBuildVersion {
+    return [System.Environment]::OSVersion.Version.Build
+}
```
I would just call this function directly instead of creating another one around it.

ryfu-msft commented on resources/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.psm1:
```diff
@@ -0,0 +1,213 @@
+    $out = [List[Language]]::new()
+
+    foreach ($locale in $localeList) {
+        $language = [Language]::new($locale.LanguageTag, $true)
```
```suggestion
        $language = [Language]::new($locale.LanguageTag)
```

nit: $true can be removed since that is the default value.

Gijsreyn replied:
This cannot be removed as it uses the constructor call.

ryfu-msft commented on resources/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.psm1:
```diff
@@ -0,0 +1,213 @@
+
+    # section to include other languages that can be installed
+    # helpful for users to discover what packages can be installed
+    $allLangues = [System.Globalization.CultureInfo]::GetCultures('AllCultures')
```
spelling: allLanguages

ryfu-msft commented on resources/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.psm1:
```diff
@@ -0,0 +1,213 @@
+    # helpful for users to discover what packages can be installed
+    $allLangues = [System.Globalization.CultureInfo]::GetCultures('AllCultures')
+    foreach ($culture in $allLangues) {
+        if ($out.LocaleName -notcontains $culture.Name -and -not ([string]::IsNullOrEmpty($culture.Name))) {
```
Does $out.LocaleName return a collection of localeNames? I feel that this will be an error. Also whether $culture.Name is null/empty should be checked first in this if statement.

ryfu-msft commented on resources/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.psm1:
```diff
@@ -0,0 +1,213 @@
+        $out.Add($language)
+    }
+
+    # section to include other languages that can be installed
+    # helpful for users to discover what packages can be installed
```
This seems to go beyound the scope of this function. I would separate this out to another funciton since my first impression was that this returns just the installed languages, not languages that can be installed.

ryfu-msft commented on resources/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.psm1:
```diff
@@ -0,0 +1,213 @@
+    [bool] Test() {
+        $currentState = $this.Get()
+
+        if ($currentState.Exist -ne $this.Exist) {
+            return $false
+        }
+        
+        return $true
```
```suggestion
     return $currentState.Exist -eq $this.Exist
```

ryfu-msft commented on resources/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.psm1:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    [Language] Get() {
+        $keyExist = [Language]::InstalledLocality.ContainsKey(($this.LocaleName))
+
+        $currentState = [Language]::InstalledLocality[$this.LocaleName]
```
You check if the key exists, then you immediately look up the value in the map. Seems like the key exists check needs to here.

ryfu-msft commented on resources/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.psm1:
```diff
@@ -0,0 +1,213 @@
+    [DscProperty()]
+    [bool] $Exist = $true
+
+    static [hashtable] $InstalledLocality
```
```suggestion
    static [hashtable] $InstalledLanguages = @{}
```

ryfu-msft commented on resources/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.psm1:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    #region Language helper functions
+    static [void] GetInstalledLocality() {   
+        [Language]::InstalledLocality = @{}
+
+        foreach ($locality in [Language]::Export()) {
+            [Language]::InstalledLocality[$locality.LocaleName] = $locality
+        }
+    }
+    #endRegion Language helper functions
```
This helper function is slightly confusing for me. Why not just use the function in export and create the installed map in the constructor?

ryfu-msft commented on resources/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.psm1:
```diff
@@ -0,0 +1,213 @@
+
+        # check if user profile contains display language
+        $userProfileLanguageDict = TryGetRegistryValue -Key (Join-Path $global:LocaleUserProfilePath $this.LocaleName) -Property 'CachedLanguageName'
+        if ((TryGetRegistryValue -Key $global:LocaleNameRegistryPath -Property $this.KeyName) -ne $this.LocaleName -and ($null -ne $userProfileLanguageDict)) {
```
I would assign TryGetRegistryValue to a variable and then use the variable in the if condition for better readability. 

Also you should check if $userProfileLanguagesDict is null first in this condition. What is $this.KeyName? I am not seeing that anywhere.

ryfu-msft commented on resources/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.psm1:
```diff
@@ -0,0 +1,213 @@
+            return $currentState
+        }
+
+        return @{
+            LocaleName = $this.LocaleName
+            Exist      = $true
+        }
```
Should this be removed?

ryfu-msft commented on resources/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.psm1:
```diff
@@ -0,0 +1,213 @@
+            return
+        }
+
+        # TODO: How do we handle sign out and sign in?
+        Set-WinUserLanguageList -Language $this.LocaleName
+
+        # TODO: Exist does not make sense here, we always want a language to exist
```
If Exist doesn't make sense here, then you don't need to use it as a parameter. Just use LocaleName as the only parameter.

ryfu-msft commented on resources/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.psm1:
```diff
@@ -0,0 +1,213 @@
+            return
+        }
+
+        # TODO: How do we handle sign out and sign in?
+        Set-WinUserLanguageList -Language $this.LocaleName
```
I would indicate to the user with a message that a reboot is required.

Gijsreyn commented:
@ryfu-msft You can put this on hold. I'm seeing some strange behaviours with the relevant modules used to set language, plus a bug found on Windows PowerShell: https://github.com/Powershell/powershellmodulecoverage/issues/18.

@denelon I might need your help, but I'll test thoroughly first and let you know.

denelon commented:
@Gijsreyn I converted this to a draft PR until you're ready for review/merge.

stephengillie commented:
Apologies for my absent repsonse - I've been too busy maintaining the [winget-pkgs](https://github.com/microsoft/winget-pkgs) repo, to be able to review this and provide feedback with constructive value.

Gijsreyn commented:
> Apologies for my absent repsonse - I've been too busy maintaining the [winget-pkgs](https://github.com/microsoft/winget-pkgs) repo, to be able to review this and provide feedback with constructive value.

Thanks for letting me know and no worries. We can't do everything at the same time.

Gijsreyn commented:
@ryfu-msft I know you're also quite busy, but if you have time to recheck this one, I would appreciate it. I incorporated as much feedback as you provided earlier. If I've missed something, then I apologize. I have rewritten many parts and tested the logic on a fresh box. Thanks in advance.

ryfu-msft commented:
/azp run

ryfu-msft commented on tests/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.Tests.ps1:
```diff
@@ -0,0 +1,81 @@
+Describe 'Language' {
+    It 'Install a preferred language' -Skip:(!$IsWindows) {
+        $desiredState = @{
+            LocaleName = 'en-GB'
```
Don't forget to change these for the ones below as well.

```suggestion
            LanguageId = 'en-GB'
```

ryfu-msft commented on resources/Microsoft.Windows.Setting.Language/Microsoft.Windows.Setting.Language.psm1:
```diff
@@ -0,0 +1,282 @@
+}
+
+function Get-LanguageList {
+    $languageList = Get-Language
```
The tests are showing that this line is causing an error. You might need to import the `LanguagePackManagement` module at the top since you have several functions that depend on that module.

ryfu-msft reviewed (changes requested):
The implementation looks good to me! I think the tests are showing that you have some module dependency issues as well as a couple lines to fix in the tests.

Gijsreyn commented:
> The implementation looks good to me! I think the tests are showing that you have some module dependency issues as well as a couple lines to fix in the tests.

Fixed up the tests. I think I was a bit to fast. Added the import module statement now. Can you give it another go?

ryfu-msft commented:
/azp run

Gijsreyn commented:
@ryfu-msft This is a bit out of my league. Clearly, the image that is being used does not have the LanguagePackmanagement built-in. I cannot find anywhere to install it. My best bet would be to either add it directly to the repository, which is quite dirty, or mock the Pester tests. Any other suggestions are welcome.

ryfu-msft commented:
> @ryfu-msft This is a bit out of my league. Clearly, the image that is being used does not have the LanguagePackmanagement built-in. I cannot find anywhere to install it. My best bet would be to either add it directly to the repository, which is quite dirty, or mock the Pester tests. Any other suggestions are welcome.
> 
> _EDIT 2: After asking the question on azure-pipelines-agent repository, I found out this module is only for clients._

Ah I see, if that is the case then you can try mocking the result of the function that uses the LanguagePackManagement so that it doesn't get invoked. The other option is the remove parts of the tests that require LanguagePackageManagement. We will attempt to have as much coverage as we can since this this is not supported on server builds.

Gijsreyn commented:
> > @ryfu-msft This is a bit out of my league. Clearly, the image that is being used does not have the LanguagePackmanagement built-in. I cannot find anywhere to install it. My best bet would be to either add it directly to the repository, which is quite dirty, or mock the Pester tests. Any other suggestions are welcome.
> > _EDIT 2: After asking the question on azure-pipelines-agent repository, I found out this module is only for clients._
> 
> Ah I see, if that is the case then you can try mocking the result of the function that uses the LanguagePackManagement so that it doesn't get invoked. The other option is the remove parts of the tests that require LanguagePackageManagement. We will attempt to have as much coverage as we can since this this is not supported on server builds.

Okay, after trying multiple approaches, I will give this one up. The reasoning being is the following:

- When mocking, it tells me that the cmdlet is not available
- Other errors that are thrown are related to a bug; see image below

![image](https://github.com/user-attachments/assets/2d07bba4-a88a-4cea-8e8f-086f56404aec)

The only thing that remains is a different OS, which changes the Azure Pipelines. I have commented out all tests, expect the validation of DSC resources. Sorry Ryan. If you have any more suggestions, they are welcome, but I don't see anyway currently to get it working on the hosted agents.
